bcda/auth/client/oktajwk.go

```diff
@@ -0,0 +1,155 @@
+package client
+
+import (
+	"crypto/rsa"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+// Gets the current set of public server signing keys. This code treats all failures as fatal,
+// since failures are likely to result in an unrecoverable state for the application.
+func getPublicKeys() (map[string]rsa.PublicKey) {
+	var (
+		body []byte
+		err  error
+		keys map[string]rsa.PublicKey
+	)
+
+	body, err = get(fmt.Sprintf("%s/oauth2/%s/.well-known/oauth-authorization-server", oktaBaseUrl, oktaServerID))
+	if err != nil {
+		logger.Errorf("can't fetch okta oauth server metadata because %v", err)
+		return nil
+	}
+	md := make(map[string]interface{})
+	if err = json.Unmarshal(body, &md); err != nil {
+		logger.Errorf("can't unmarshal %s because %v", string(body), err)
+		return nil
+	}
+
+	jwkUrl := md["jwks_uri"].(string)
+	body, err = get(jwkUrl)
+	if err != nil {
+		logger.Errorf("can't get keys from %s because %v", jwkUrl, err)
+		return nil
+	}
+
+	keys, err = parseKeys(body)
+	if err != nil {
+		logger.Errorf("couldn't parse %s because %v", string(body), err)
+		return nil
+	}
+	if len(keys) == 0 {
+		logger.Errorf("must have at least 1 public key; have none")
+		return nil
+	}
+
+	logger.Infof("%d okta public oauth server public keys cached", len(keys))
+	return keys
+}
+
+type RsaJWK struct {
+	KeyType   string `json:"kty"`
+	Algorithm string `json:"alg"`
+	ID        string `json:"kid"`
+	Use       string `json:"use"`
+	E         string `json:"e"`
+	N         string `json:"n"`
+}
+
+type KeyList struct {
+	Keys []*RsaJWK `json:"keys"`
+}
+
+func get(urlString string) ([]byte, error) {
+	if urlString == "" {
+		err := fmt.Errorf("a non-empty okta metadata url must be provided; is an env var not set")
+		logger.Error(err)
+		return nil, err
+	}
+	u, err := url.Parse(urlString)
+	if err != nil {
+		logger.Errorf("invalid url %s because %v", urlString, err)
+		return nil, err
+	}
+
+	var client = &http.Client{Timeout: time.Second * 10,}
+	res, err := client.Get(u.String())
+	if err != nil {
+		logger.Errorf("net failure to %s because %v", u.String(), err)
```

Comment:
Having chosen to use structured logging, we should be putting our common arguments in separate fields.  And we should standardize the field names.  Out of the scope of this PR would be a logging library to help with that.  But in the meantime, try:

`logger.WithFields(logrus.Fields{"error": err, "url": u.String()}).Error("net failure in get")`

NB: don't use more than 2 `.WithFields()` calls on a single logrus call.